src/index.ts

```diff
@@ -476,138 +661,234 @@ export class Upload extends Pumpify {
 
-    this.setPipeline(bufferStream, offsetStream, delayStream);
+        const result = await upstreamQueue.next();
 
-    this.pipe(requestStreamEmbeddedStream);
+        if (result.value) {
+          this.numChunksReadInRequest++;
+          this.lastChunkSent = result.value.chunk;
+          this.numBytesWritten += result.value.chunk.byteLength;
 
-    this.once('restart', () => {
-      // The upload is being re-attempted. Disconnect the request
-      // stream, so it won't receive more data.
-      this.unpipe(requestStreamEmbeddedStream);
+          this.emit('progress', {
+            bytesWritten: this.numBytesWritten,
+            contentLength: this.contentLength,
+          });
+
+          requestStream.push(result.value.chunk, result.value.encoding);
+        }
+
+        if (result.done) {
+          requestStream.push(null);
+        }
+      },
     });
 
+    // This should be 'once' as `startUploading` can be called again for
+    // multi chunk uploads and each request would have its own response.
+    this.once('response', resp => {
+      responseReceived = true;
+      this.responseHandler(resp);
+    });
+    let headers: GaxiosOptions['headers'] = {};
+
+    // If using multiple chunk upload, set appropriate header
+    if (multiChunkMode && expectedUploadSize) {
+      // The '-1' is because the ending byte is inclusive in the request.
+      const endingByte = expectedUploadSize + this.numBytesWritten - 1;
+      headers = {
+        'Content-Length': expectedUploadSize,
+        'Content-Range': `bytes ${this.offset}-${endingByte}/${this.contentLength}`,
+      };
+    } else {
+      headers = {
+        'Content-Range': `bytes ${this.offset}-*/${this.contentLength}`,
+      };
+    }
+
     const reqOpts: GaxiosOptions = {
       method: 'PUT',
       url: this.uri,
-      headers: {
-        'Content-Range': 'bytes ' + this.offset + '-*/' + this.contentLength,
-      },
-      body: requestStreamEmbeddedStream,
+      headers,
+      body: requestStream,
     };
 
     try {
       await this.makeRequestStream(reqOpts);
-    } catch (e) {
+    } catch (err) {
+      const e = err as Error;
       this.destroy(e);
     }
   }
 
-  private onChunk(
-    chunk: string,
-    enc: BufferEncoding,
-    next: (err?: Error, data?: string) => void
-  ) {
-    const offset = this.offset!;
-    const numBytesWritten = this.numBytesWritten;
+  // Process the API response to look for errors that came in
+  // the response body.
+  private responseHandler(resp: GaxiosResponse) {
+    if (resp.data.error) {
+      this.destroy(resp.data.error);
+      return;
+    }
 
-    this.emit('progress', {
-      bytesWritten: this.numBytesWritten,
-      contentLength: this.contentLength,
-    });
+    const shouldContinueWithNextMultiChunkRequest =
+      this.chunkSize &&
+      resp.status === RESUMABLE_INCOMPLETE_STATUS_CODE &&
+      resp.headers.range;
+
+    if (shouldContinueWithNextMultiChunkRequest) {
```

Comment:
There are a few places like this, where we have almost completely different logic for chunk upload, vs., regular upload, but share a method.

It makes me wonder if the implementation would read slightly cleaner if we had, `export class Upload extends Pumpify` in `index.ts`, as it is today, but then introduced an additional class, `export class ChunkUpload extends Upload`.

It might introduce a bit of repeated logic, but I'm curious if it would make the implementation even more elegant.

Reply:
+1, I think separating out the logic more between the two upload types would definitely make the code easier to read.

Reply:
Good point - I figured after the refactor to support multiple requests most of the logic would be the same with the exception of 1 parameter (`upstreamIterator` in `startUploading`) and 3 `if` statements (including this one).

I was hoping `shouldContinueWithNextMultiChunkRequest` would make it easier to understand what's happening, but seemed to have backfired.

Reply:
As I said, not blocking for me, but if you have the time to experiment with the refactor, I'd be curoius what it looks like.